library/Mockery/Exception/NoMatchingExpectationException.php

```diff
@@ -20,17 +20,84 @@
 
 namespace Mockery\Exception;
 
+use Hamcrest\Util;
 use Mockery;
+use SebastianBergmann\Comparator\ComparisonFailure;
+use SebastianBergmann\Comparator\Factory;
 
 class NoMatchingExpectationException extends Mockery\Exception
 {
-    protected $method = null;
+    /**
+     * @var string
+     */
+    protected $method;
 
-    protected $actual = array();
+    /**
+     * @var array
+     */
+    protected $actual;
 
-    protected $mockObject = null;
+    /**
+     * @var Mockery\MockInterface
+     */
+    protected $mockObject;
 
-    public function setMock(Mockery\LegacyMockInterface $mock)
+    /**
+     * @param string $methodName
+     * @param array $actualArguments
+     * @param array $expectations
+     */
+    public function __construct(
+        Mockery\MockInterface $mock,
+        $methodName,
+        $actualArguments,
+        $expectations
+    ) {
+        $this->setMock($mock);
+        $this->setMethodName($methodName);
+        $this->setActualArguments($actualArguments);
+
+        $diffs = [];
+        foreach ($expectations as $expectation) {
+            $expectedArguments = $expectation->getExpectedArgs();
+
+            $diff = $this->diff(
+                $this->normalizeForDiff($expectedArguments),
+                $this->normalizeForDiff($actualArguments)
+            );
+            if (null === $diff) {
+                // If we reach this, it means that the exception has not been
+                // raised by a non-strict equality. So the diff is null.
+                // We do the comparison again but this time comparing references
+                // of objects.
+                $diff = $this->diff(
+                    $this->normalizeForStrictDiff($expectedArguments),
+                    $this->normalizeForStrictDiff($actualArguments)
+                );
+            }
+
+            $diffs[] = sprintf(
+                "\n%s::%s with arguments%s",
+                $expectation->getMock()->mockery_getName(),
+                $expectation->getName(),
+                null !== $diff ? $diff : "\n### No diff ###"
+            );
+        }
+
+        $message = 'No matching expectation found for '
+            . $this->getMockName() . '::'
+            . \Mockery::formatArgs($methodName, $actualArguments)
+            . '. Either the method was unexpected or its arguments matched'
+            . ' no expected argument list for this method.'
+            . PHP_EOL . PHP_EOL
+            . 'Here is the list of available expectations and their diff with actual input:'
+            . PHP_EOL
+            . implode('', $diffs);
+
+        parent::__construct($message, 0, null);
+    }
+
+    public function setMock(Mockery\MockInterface $mock)
```

Comment:
Is this required to be a public method? Can it be private?

Reply:
You are right. Changed for all public methods in this class.